### ISSUE TITLE
Crash after Windows 10: 2004 update

### ISSUE THREAD

**Issue by Houligan86** [PrivateServerEmulator/ME3PSE#6]
In the server, I get a redirector crash
win32exception: the client and server cannot communicate, because they do not possess a common algorithm
I am guessing whatever protocol ME3 uses is disabled in the new update.

The MEtweaker forums require registration, which is currently disabled, so I can't ask there.

**Comment by Erik-JS** [comment]
The ME3Tweaks forum is read-only now, but the [PSE FAQ](https://me3tweaks.com/forums/viewtopic.php?f=10&t=9) is still available.

Another PSE user told me on Discord that the tool I made to add old cipher suites to the list of supported suites didn't work on the 2004 update.

Did you try the PowerShell method described in the [FAQ](https://me3tweaks.com/forums/viewtopic.php?f=10&t=9)?

`Enable-TlsCipherSuite -Name 'TLS_RSA_WITH_RC4_128_SHA'`
`Enable-TlsCipherSuite -Name 'TLS_RSA_WITH_RC4_128_MD5'`
(must run with admin rights)

Does this fail too?

**Comment by Houligan86** [comment]
Sorry for the late reply. I went through those steps and it still didn't work. On this computer, it was working until the 2004 (may) update, which was installed right before I had this issue.

My setup is I have two computers, one to run the server and one to play the game on. If I run the server from the gaming computer, I and my friends are able to connect, using Hamachi to make a LAN.

I have access to visual studio if there is additional debugging information I can collect using it.

I also disabled the firewall while I tried this to eliminate that as a cause.

**Comment by Erik-JS** [comment]
I made another tool to add old cipher suites back to the list of allowed cipher suites. It uses Windows API directly, the other tool was calling PowerShell in background.

[CipherSuiteTest.zip](https://github.com/PrivateServerEmulator/ME3PSE/files/4907302/CipherSuiteTest.zip)
Source code: https://github.com/Erik-JS/CipherSuiteTest

**Comment by Houligan86** [comment]
That tool returns that the ciphers are present.

**Comment by Erik-JS** [comment]
I set up a VM with Windows 10 v2004 and now I see what's going on. Old cipher suites can be enabled, but PSE acts like they're missing.

Looks like v2004 is forcibly blocking the use of old cipher suites or there's something else missing here that I have no idea of what could it be at the moment.

**Comment by Houligan86** [comment]
I think I figured it out. I did the reverse of these instructions (https://www.digicert.com/kb/ssl-support/iis-disabling-ssl-v3.htm).
Regedit, navigate to HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\Schannel\
Add Key "Protocols"
Add Key "SSL 3.0"
Add Keys "Client" and "Server"
Under Client, New DWORD (32-bit), named "DisabledByDefault", value 0
Under Server, New DWORD (32-bit), named "Enabled", value 1

**Comment by Erik-JS** [comment]
Good. I found [this article](https://techcommunity.microsoft.com/t5/core-infrastructure-and-security/demystifying-schannel/ba-p/259233) which mentions those reg paths.

From what I tested here with VirtualBox (host: 1903, guest: 2004), only "Server" needs to have "Enabled = 1", and you don't need to restart Windows after adding, changing or deleting this setting on registry.

It seems that SSL3 server functionality is now "disabled unless stated otherwise". According to [this](https://docs.microsoft.com/pt-br/windows/win32/secauthn/protocols-in-tls-ssl--schannel-ssp-), SSL3 server support has been disabled since 1607... I don't know if that's accurate because [when we started having problems with cipher suites in 1709](https://me3tweaks.com/forums/viewtopic.php?f=10&t=580#p1762) we never had to change anything in registry regarding SSL3 server state.

I'll add cipher suite checking and SSL3 server state verification (only for Windows 10 2004 or above) to PSE and make a new release when done.

@Houligan86: let's keep this issue open for now so other people can be aware of this. I'll close it when the next release comes.

**Anyone else reading this:** grab the CipherSuiteTest tool posted above to check and add required cipher suites. Also, grab this reg file to enable SSL3 server functionality on Windows 10 v2004 or later:
[SSL3_enable-server.zip](https://github.com/PrivateServerEmulator/ME3PSE/files/4908593/SSL3_enable-server.zip)

If the cipher suite issue is anything to go by, then SSL3 server will get "disabled" again when the next Windows 10 update comes. I believe this should be considered a red flag for the end of SSL3 support under Schannel - I'm already looking for alternatives.
